02_Util/src/security/SignedMeterValuesUtil.ts

```diff
@@ -0,0 +1,222 @@
+   *
+   * If a meter value is unsigned, it is valid.
+   *
+   * If a meter value is signed, it is valid if:
+   * - SignedMeterValuesConfig is configured
+   * AND
+   * - The incoming signed meter value's signing method matches the configured signing method
+   * AND
+   * - The incoming signed meter value's public key is empty but there is a public key stored for that charging station
+   * OR
+   * - The incoming signed meter value's public key isn't empty and it matches the configured public key
+   *
+   * @param stationId - The charging station the meter values belong to
+   * @param meterValues - The list of meter values
+   */
+  public async validateMeterValues(
+    stationId: string,
+    meterValues: [MeterValueType, ...MeterValueType[]],
+  ): Promise<boolean> {
+    let validMeterValues = true;
+
+    for (const meterValue of meterValues) {
+      for (const sampledValue of meterValue.sampledValue) {
+        if (sampledValue.signedMeterValue) {
+          validMeterValues = await this.validateSignedSampledValue(
+            stationId,
+            sampledValue.signedMeterValue,
+          );
+          if (!validMeterValues) {
+            return validMeterValues;
+          }
+        }
+      }
+    }
+
+    return validMeterValues;
+  }
+
+  private async validateSignedSampledValue(
+    stationId: string,
+    signedMeterValue: SignedMeterValueType,
+  ): Promise<boolean> {
+    if (signedMeterValue.publicKey && signedMeterValue.publicKey.length > 0) {
+      const incomingPublicKeyIsValid =
+        await this.validateSignedMeterValueSignature(signedMeterValue);
+
+      if (this._signedMeterValuesConfiguration && incomingPublicKeyIsValid) {
+        await this._chargingStationSecurityInfoRepository.readOrCreateChargingStationInfo(
+          stationId,
+          this._signedMeterValuesConfiguration.publicKeyFileId,
+        );
+
+        return true;
+      } else {
+        return false;
+      }
+    } else {
+      const chargingStationPublicKeyFileId =
+        await this._chargingStationSecurityInfoRepository.readChargingStationPublicKeyFileId(
+          stationId,
+        );
+      return await this.validateSignedMeterValueSignature(
+        signedMeterValue,
+        chargingStationPublicKeyFileId,
+      );
+    }
+  }
+
+  private async validateSignedMeterValueSignature(
+    signedMeterValue: SignedMeterValueType,
+    publicKeyFileId?: string,
+  ): Promise<boolean> {
+    const incomingPublicKeyString = signedMeterValue.publicKey;
+    const signingMethod = signedMeterValue.signingMethod;
+
+    if (
+      !this._signedMeterValuesConfiguration?.publicKeyFileId ||
+      (publicKeyFileId &&
+        publicKeyFileId !==
+          this._signedMeterValuesConfiguration?.publicKeyFileId) ||
+      (!publicKeyFileId && incomingPublicKeyString.length === 0) ||
+      this._signedMeterValuesConfiguration?.signingMethod !== signingMethod
+    ) {
+      return false;
+    }
```

Comment:
Would maybe suggest breaking this if statement up so that we can log out exactly why the signature is invalid:

```suggestion
    if (!this._signedMeterValuesConfiguration?.publicKeyFileId) {
      console.log("Invalid signature due to public key missing in system config");
      return false;
    }
    if (!publicKeyFileId) {
      console.log("Invalid signature due to public key missing from request");
      return false;
    }
    if (publicKeyFileId !== this._signedMeterValuesConfiguration?.publicKeyFileId)) {
      console.log("Invalid signature due to public key not matching system config public key");
      return false;    
    }
```

Reply:
very true -- updated so that there are more explicit blocks with logging 👍